src/Compiler/Microsoft.CodeAnalysis.Razor/src/SymbolExtensions.cs

```diff
@@ -13,33 +14,76 @@ internal static class SymbolExtensions
             .RemoveMiscellaneousOptions(SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
 
     /// <summary>
-    /// Checks if <paramref name="symbol"/> has the same fully qualified name as <paramref name="fullName"/>.
+    /// Checks if <paramref name="typeSymbol"/> has the same fully qualified name as <paramref name="fullName"/>.
     /// </summary>
     /// <remarks>
     /// This is a *very* simple implementation that doesn't handle all cases. It's only intended to be used
     /// for a small set of known types. If you add a new type to be looked up by this function you must ensure
-    /// the simple check works correctly for it. 
+    /// the simple check works correctly for it.
     /// </remarks>
     /// <returns>true if the type has the specified full name</returns>
-    internal static bool HasFullName(this ISymbol symbol, string fullName)
+    internal static bool HasFullName(this ITypeSymbol typeSymbol, string fullName)
     {
         // generally we expect this call to fail, so try and match on the simple name first, before
         // checking the fully qualified name, to prevent unnecessary string allocation from ToDisplayString 
-        if (symbol.Name.Length > fullName.Length)
+        if (typeSymbol.Name.Length > fullName.Length)
         {
             return false;
         }
 
-        // the simple name doesn't include generic type params, so skip over them in the full name if they are present
-        var fullNameEndIndex = fullName[^1] == '>' ? fullName.LastIndexOf('<') : fullName.Length;
-        var fullNameStartIndex = fullName.LastIndexOf('.', fullNameEndIndex - 1) + 1;
+        var fullNameSpan = fullName.AsSpan();
 
-        if (!fullName.AsSpan(fullNameStartIndex, fullNameEndIndex - fullNameStartIndex).Equals(symbol.Name.AsSpan(), StringComparison.Ordinal))
+        // The simple name doesn't include generic type params, so skip over them in the full name if they are present.
+        // UNDONE: This will need adjustment if this method should need to handle types overloaded on the number
+        // of generic type parameters.
+        if (fullNameSpan[^1] == '>')
         {
-            return false;
+            var endIndex = fullNameSpan.LastIndexOf('<');
```

Comment:
>var endIndex = fullNameSpan.LastIndexOf('<'); [](http://example.com/codeflow?start=12&length=45)

Could this ever have multiple of '<'?

Reply:
Good catch! Although, this is how the method behaved before my method. It's a quick-and-dirty function that is only expected to work with the relatively small number of inputs that the Razor compiler feeds it.